tensorboard/data/server/gcs/auth.rs

```diff
@@ -235,6 +242,42 @@ impl Credentials {
     }
 }
 
+/// Partial structure of the `GOOGLE_APPLICATION_CREDENTIALS` file.
+#[derive(Deserialize)]
+struct ApplicationCreds {
+    r#type: Option<String>,
+    client_id: Option<String>,
+    client_secret: Option<String>,
+    refresh_token: Option<String>,
+}
+
+/// User's credentials may be valid, but are not supported.
+///
+/// Currently, we only support OAuth refresh tokens, not service account private keys.
+#[derive(Debug, thiserror::Error)]
+#[error("unsupported credentials of type {:?}; only OAuth refresh tokens supported", .creds_type)]
+pub struct UnsupportedCredentialsError {
+    /// The `"type"` field found in the credentials JSON file, for informational purposes.
+    creds_type: String,
+}
+
+impl ApplicationCreds {
+    fn into_credentials(self) -> Result<Credentials, UnsupportedCredentialsError> {
+        match (self.client_id, self.client_secret, self.refresh_token) {
+            (Some(client_id), Some(client_secret), Some(refresh_token)) => {
+                Ok(Credentials::RefreshToken(RefreshToken(RefreshTokenCreds {
+                    client_id,
+                    client_secret,
+                    refresh_token,
+                })))
+            }
+            _ => Err(UnsupportedCredentialsError {
+                creds_type: self.r#type.unwrap_or_default(),
+            }),
```

Comment:
I am already forgetting a lot of Rust so help me understand here: so we get UnsupportedCredentialsError when the JSON does not contain all of `client_id`,`client_secret`, and `refresh_token`, right? In which scenario would we have that?  

If I am reading the description correctly, service account would fall into this case? In such case, I am not 100% sure if just printing `type` (which can be missing, too) is actionable/comprehendible. 

Did I understand the flow correctly?

Reply:
You understand correctly. A refresh token creds file looks like:

```json
{
  "type": "authorized_user",
  "client_id": "...",
  "client_secret": "...",
  "refresh_token": "..."
}
```

and a service account creds file looks like:

```json5
{
  "type": "service_account",
  "project_id": "<GCP project ID>",
  "private_key": "...",
  /* also some more string fields here (but not `refresh_token`, etc.) */
}
```

I suspect that `type` should always be there; I just made it optional
because I couldn’t find actual docs/specs for this file format, so
wanted to be a bit conservative.

Note that we also print the file name. So the actual error message, on a
real service account file, is:

> unsupported GCS credentials of type "service_account"; only OAuth refresh tokens supported (credentials file /tmp/my-service-account-key.json)

I think that this is pretty readable. WDYT?

Reply:
I see, so there is some implicit trust that the parsable JSON is correctly formed like `type: service_account` always misses `refresh_token` and etc... 

Not sure if I want to trust random file on the file system but probability of it being corrupted is low, I guess.

LGTM.

Reply:
Yeah—I think the failure modes here are all pretty tolerable. Even if
you have a somehow-corrupted credentials file, we’ll emit a clear error
and exit cleanly (or, with #4794, fall back to `--load_fast=false`).
And if something goes wrong, we just fail to authenticate; no
possibility of losing, misrepresenting, or corrupting the actual data.

In case it helps, the [TensorFlow logic][1] is similarly ad hoc. It just
assumes that if there’s a `refresh_token` field, the client ID and
secret must also be there, and likewise that if there’s a `private_key`
field then `private_key_id`, et al. must be present.

[1]: https://github.com/tensorflow/tensorflow/blob/da86421df896248401341c9f1a6b23cefb267a3a/tensorflow/core/platform/cloud/google_auth_provider.cc#L219-L234